File: backend/app/schemas/atomic_metric.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class AtomicMetricQueryRequest(BaseModel):
    model_config = ConfigDict(extra="ignore", populate_by_name=True)

    data_source: str | None = None
    dept_code: str | None = None
    date_start: str | None = None
    date_end: str | None = None
    document_type: str | None = None
    category_code: str | None = None
    category_name: str | None = None
    type_code: str | None = None
    subtype_code: str | None = None
    include_yoy: bool = False
    include_mom: bool = False
    include_share: bool = False
    include_mom_count: bool = False
    include_yoy_count: bool = False
    include_cumulative: bool = False
    include_dim_combo: bool = False
    dim_combo_levels: str | None = None
    include_category_share: bool = False
    include_type_share: bool = False
    include_subtype_share: bool = False
    include_hot_community: bool = False
    org_dimension: str | None = None
    include_hot_period: bool = False
    hot_period_hours: int | None = None
    include_region_table: bool = False
    filter_duplicate: bool = False
    exclude_non_police: bool = False
    exclude_traffic: bool = False
    filter_self_received: bool = False
    exclude_self_received: bool = False
    tag_package_id: int | None = None
    yoy_trend: str | None = None
    trend_compare: str | None = None
    yoy_analysis_drill: str | None = None
    yoy_trend_top_n: int | None = None
    rank_sort_by: str | None = None
    rank_sort_order: str | None = None
    count_threshold_op: str | None = None
    count_threshold_value: float | None = None
    include_warning: bool = False
    warning_rule_type: str | None = None
    params: dict[str, Any] = Field(default_factory=dict)


class AtomicMetricQueryResult(BaseModel):
    model_config = ConfigDict(extra="ignore")

    total: int | float | None = None
    yoy: float | None = None
    mom: float | None = None
    yoy_change: str | None = None
    mom_change: str | None = None
    yoy_count: int | float | None = None
    mom_count: int | float | None = None
    cumulative: int | float | None = None
    dim_combo: str | None = None
    dim_table_headers: list[str] | None = None
    dim_table_rows: list[list[str]] | None = None
    share: float | None = None
    category_share: str | None = None
    type_share: str | None = None
    subtype_share: str | None = None
    hot_communities: str | None = None
    org_units: str | None = None
    hot_periods: str | None = None
    regions: str | None = None
    yoy_stations: str | None = None
    warning_text: str | None = None
    table_title: str | None = None
    table_headers: list[str] | None = None
    table_rows: list[list[str]] | None = None
    html_fragment: str | None = None
    field_values: dict[str, Any] = Field(default_factory=dict)
    content_segments: list[dict[str, Any]] = Field(default_factory=list)
    text_content: str | None = None
    executed_sql: str | None = None

File: frontend/src/api/atomicMetric.ts
import { apiPost } from '@/api/request'

/** 与后端 AtomicMetricQueryRequest 对齐（snake_case） */
export interface AtomicMetricQueryPayload {
  data_source?: string | null
  dept_code?: string | null
  date_start?: string | null
  date_end?: string | null
  document_type?: string | null
  category_code?: string | null
  category_name?: string | null
  type_code?: string | null
  subtype_code?: string | null
  include_yoy?: boolean
  include_mom?: boolean
  include_share?: boolean
  include_mom_count?: boolean
  include_yoy_count?: boolean
  include_cumulative?: boolean
  include_dim_combo?: boolean
  dim_combo_levels?: string | null
  include_category_share?: boolean
  include_type_share?: boolean
  include_subtype_share?: boolean
  include_hot_community?: boolean
  org_dimension?: string | null
  include_hot_period?: boolean
  hot_period_hours?: number | null
  include_region_table?: boolean
  filter_duplicate?: boolean
  exclude_non_police?: boolean
  exclude_traffic?: boolean
  filter_self_received?: boolean
  exclude_self_received?: boolean
  tag_package_id?: number | null
  yoy_trend?: string | null
  trend_compare?: string | null
  yoy_analysis_drill?: string | null
  yoy_trend_top_n?: number | null
  rank_sort_by?: string | null
  rank_sort_order?: string | null
  count_threshold_op?: string | null
  count_threshold_value?: number | null
  include_warning?: boolean
  warning_rule_type?: string | null
  params?: Record<string, unknown>
}

/** 与后端 AtomicMetricQueryResult 对齐（snake_case） */
export interface AtomicMetricQueryResult {
  total?: number | null
  yoy?: number | null
  mom?: number | null
  yoy_change?: string | null
  mom_change?: string | null
  yoy_count?: number | null
  mom_count?: number | null
  cumulative?: number | null
  dim_combo?: string | null
  dim_table_headers?: string[] | null
  dim_table_rows?: string[][] | null
  share?: number | null
  category_share?: string | null
  type_share?: string | null
  subtype_share?: string | null
  hot_communities?: string | null
  org_units?: string | null
  hot_periods?: string | null
  regions?: string | null
  yoy_stations?: string | null
  warning_text?: string | null
  table_title?: string | null
  table_headers?: string[] | null
  table_rows?: string[][] | null
  html_fragment?: string | null
  field_values?: Record<string, unknown>
  content_segments?: Array<Record<string, unknown>>
  text_content?: string | null
  executed_sql?: string | null
}

export function queryAtomicMetric(
  data: AtomicMetricQueryPayload,
  options?: { signal?: AbortSignal; timeout?: number },
) {
  return apiPost<AtomicMetricQueryResult>('/atomic-metric/query', data, {
    timeout: 180000,
    ...options,
  })
}
